mapper/orm/src/main/java/org/hibernate/search/mapper/orm/batchindexing/impl/BatchCoordinator.java

```diff
@@ -0,0 +1,174 @@
+import java.util.concurrent.Future;
+
+import org.hibernate.CacheMode;
+import org.hibernate.engine.spi.SessionFactoryImplementor;
+import org.hibernate.search.mapper.orm.logging.impl.Log;
+import org.hibernate.search.mapper.orm.mapping.spi.HibernateOrmMapping;
+import org.hibernate.search.mapper.pojo.work.spi.PojoWorkPlan;
+import org.hibernate.search.util.AssertionFailure;
+import org.hibernate.search.util.impl.common.Executors;
+import org.hibernate.search.util.impl.common.LoggerFactory;
+import java.lang.invoke.MethodHandles;
+
+/**
+ * Makes sure that several different BatchIndexingWorkspace(s)
+ * can be started concurrently, sharing the same batch-backend
+ * and IndexWriters.
+ *
+ * @author Sanne Grinovero
+ */
+public class BatchCoordinator extends ErrorHandledRunnable {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	private final Set<Class<?>> rootEntities; //entity types to reindex excluding all subtypes of each-other
+	private final SessionFactoryImplementor sessionFactory;
+	private final HibernateOrmMapping mapping;
+
+	//TODO: use the workPlan to handle optimize tasks
+	private final PojoWorkPlan workPlan;
+
+	private final int typesToIndexInParallel;
+	private final int documentBuilderThreads;
+	private final CacheMode cacheMode;
+	private final int objectLoadingBatchSize;
+	private final boolean optimizeAtEnd;
+	private final boolean purgeAtStart;
+	private final boolean optimizeAfterPurge;
+	private final CountDownLatch endAllSignal;
+	private final long objectsLimit;
+	private final int idFetchSize;
+	private final Integer transactionTimeout;
+	private final String tenantId;
+	private final List<Future<?>> indexingTasks = new ArrayList<>();
+
+	public BatchCoordinator(Set<Class<?>> rootEntities,
+							SessionFactoryImplementor sessionFactory,
+							HibernateOrmMapping mapping,
+							PojoWorkPlan workPlan,
+							int typesToIndexInParallel,
+							int documentBuilderThreads,
+							CacheMode cacheMode,
+							int objectLoadingBatchSize,
+							long objectsLimit,
+							boolean optimizeAtEnd,
+							boolean purgeAtStart,
+							boolean optimizeAfterPurge,
+							int idFetchSize,
+							Integer transactionTimeout,
+							String tenantId) {
+		this.idFetchSize = idFetchSize;
+		this.transactionTimeout = transactionTimeout;
+		this.tenantId = tenantId;
+		this.rootEntities = rootEntities;
+		this.sessionFactory = sessionFactory;
+		this.mapping = mapping;
+		this.workPlan = workPlan;
+		this.typesToIndexInParallel = typesToIndexInParallel;
+		this.documentBuilderThreads = documentBuilderThreads;
+		this.cacheMode = cacheMode;
+		this.objectLoadingBatchSize = objectLoadingBatchSize;
+		this.optimizeAtEnd = optimizeAtEnd;
+		this.purgeAtStart = purgeAtStart;
+		this.optimizeAfterPurge = optimizeAfterPurge;
+		this.objectsLimit = objectsLimit;
+		this.endAllSignal = new CountDownLatch( rootEntities.size() );
+	}
+
+	@Override
+	public void runWithErrorHandler() {
```

Comment:
Good call removing the `BatchBackend` here. It doesn't make much sense in Search 6.